### DISCUSSION TITLE
Is smart expand/shrink selection supported?

### DISCUSSION THREAD

**Discussion by rynoV** [ionide/ionide-vscode-fsharp#1982] [Q&A]
Referring to the [shrink/expand selection](https://code.visualstudio.com/docs/editor/codebasics#_shrinkexpand-selection) feature in vscode, does ionide support this selection based on syntax? For example expanding selection to the expression at point, or current let binding

I was looking at [this commit](https://github.com/fsharp/FsAutoComplete/commit/cf76ab63f28c5854fdd8e815e14593e1a6424d7f) which makes me think it should be supported, but for me the selection works essentially like a text file, only expanding based on whitespace.

**Comment by baronfel** [comment]
It is supported, but it's based on the untyped walker that you see in that commit. If there are shrinks/expands that feel weird, it's possible that we're not walking the tree correctly for that syntax shape, or some other similar problem in our implementation.

**Reply by rynoV** [reply]
Thanks for the quick reply!

> If there are shrinks/expands that feel weird, it's possible that we're not walking the tree correctly for that syntax shape, or some other similar problem in our implementation.

It seems more that the feature isn't even enabled, and it's just falling back on the default whitespace-based expansion: word, inner line, full line, file.

If it helps, this is what I have in the "Ionide" part of the "Output" panel:

```
[11:17:11 DEBUG] [LanguageService] FSAC (NETCORE): '/home/calum/.vscode/extensions/ionide.ionide-fsharp-7.17.0/bin/net7.0/fsautocomplete.dll'
[11:17:11 DEBUG] [LanguageService] F# language server options: % {
  command: '/usr/bin/dotnet',
  args: [
    '/home/calum/.vscode/extensions/ionide.ionide-fsharp-7.17.0/bin/net7.0/fsautocomplete.dll',
    '--project-graph-enabled',
    '--state-directory',
    '/home/calum/.config/Code/User/workspaceStorage/c2aebd8e4759628d6dec0e971b0dd589/Ionide.Ionide-fsharp'
  ],
...env
```

And this is the "F#" output after restarting the language service and making a change to the file:

```
[Error - 12:12:16] Request textDocument/inlayHint failed.
  Message: No typecheck results for <filepath>.fs
  Code: -32603 
[Error - 12:12:16] Request textDocument/documentSymbol failed.
  Message: No declarations for <filepath>.fs
  Code: -32603 
[Error - 12:12:16] Request textDocument/semanticTokens/range failed.
  Message: No typecheck results for <filepath>.fs
  Code: -32603 
[Error - 12:12:16] Request textDocument/semanticTokens/range failed.
  Message: No typecheck results for <filepath>.fs
  Code: -32603 
[Error - 12:12:16] Request textDocument/semanticTokens/range failed.
  Message: No typecheck results for <filepath>.fs
  Code: -32603 
[Error - 12:12:16] Request textDocument/semanticTokens/range failed.
  Message: No typecheck results for <filepath>.fs
  Code: -32603 
[12:13:00.534 ERR] [FsAutoComplete.Lsp.AdaptiveFSharpLspServer] TextDocumentDidChange Request Errored {"TextDocument": {"Uri": "file://<filepath>.fs", "Version": 45, "$type": "VersionedTextDocumentIdentifier"}, "ContentChanges": [{"Range": {"Value": {"Start": {"Line": 508, "Character": 54, "$type": "Position"}, "End": {"Line": 508, "Character": 55, "$type": "Position"}, "$type": "Range"}, "$type": "Some"}, "RangeLength": {"Value": 1, "$type": "Some"}, "Text": "", "$type": "TextDocumentContentChangeEvent"}], "$type": "DidChangeTextDocumentParams"}
System.Threading.Tasks.TaskCanceledException: A task was canceled.
```

The `textDocument/semanticTokens/range` error seems relevant, although it's weird because browsing document symbols with Ctrl+Shift+O works fine even though there was an error for that

**Reply by baronfel** [reply]
This mostly seems like the language server hasn't correctly typechecked your code, and this specific feature is coded against typechecking, not just syntax trees, so you're getting that error. In the absence of data coming back from the LSP, VSCode is falling back to the 'tokenization' rules that we tell it about F#.

**Reply by rynoV** [reply]
I tried the same with the `FirstIonideProject` from [here](https://learn.microsoft.com/en-us/dotnet/fsharp/get-started/get-started-vscode), pretty much the same result:

```
[Error - 12:25:20] Request textDocument/documentSymbol failed.
  Message: No declarations for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:25:20] Request textDocument/documentSymbol failed.
  Message: No declarations for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:25:20] Request textDocument/inlayHint failed.
  Message: No typecheck results for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:25:20] Request textDocument/semanticTokens/range failed.
  Message: No typecheck results for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:25:20] Request textDocument/hover failed.
  Message: No typecheck results for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:25:20] Request textDocument/semanticTokens/full failed.
  Message: No typecheck results for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
```

In the Program.fs file below, with cursor in the string, it selects by word then line (but I'm expecting it to select within and around the string)

```f#
// For more information see https://aka.ms/fsharp-console-apps
printfn "Hello from F#"
```

**Reply by rynoV** [reply]
> This mostly seems like the language server hasn't correctly typechecked your code, and this specific feature is coded against typechecking, not just syntax trees, so you're getting that error. In the absence of data coming back from the LSP, VSCode is falling back to the 'tokenization' rules that we tell it about F#.

Hmm I wonder if the order of typechecking and other operations is off then, because both projects are typechecking otherwise and other language features are working

**Reply by baronfel** [reply]
the tooling still hasn't typechecked, so I wouldn't expect any difference. Can you try doing an explicit `dotnet restore` for the project and then reload VSCode?

**Reply by rynoV** [reply]
I closed vscode, ran `dotnet restore` in `FirstIonideProject`, after the first time reopening vscode it had the same behaviour and there wasn't any "F#" in the "Output" panel, I repeated the process and now there is "F#" with pretty much the same output and same behaviour:

```
[Error - 12:37:37] Request textDocument/documentSymbol failed.
  Message: No declarations for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:37:37] Request textDocument/documentSymbol failed.
  Message: No declarations for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:37:37] Request textDocument/inlayHint failed.
  Message: No typecheck results for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:37:37] Request textDocument/semanticTokens/range failed.
  Message: No typecheck results for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
[Error - 12:37:37] Request textDocument/semanticTokens/full failed.
  Message: No typecheck results for /home/calum/code/FirstIonideProject/Program.fs
  Code: -32603 
```

**Reply by rynoV** [reply]
It also seems to be targeting net8.0 for FirstIonideProject, if that makes any difference. And for my original project it's targeting net6.0, and there is multiple `fsproj` files in subdirectories. On the original project `dotnet restore` also didn't change anything, and opening vscode only in one of the subdirectories didn't change anything either.
